File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    void findCombination(int ind,int target,int [] candidates,  List<List<Integer>> ans,List<Integer> ds){
        if(ind==candidates.length){
            if(target==0){
                ans.add(new ArrayList<>(ds));
            }
            return ;
        }
        if(candidates[ind]<=target){
            ds.add(candidates[ind]);
            findCombination(ind,target-candidates[ind],candidates,ans,ds);
            ds.remove(ds.size() - 1);
        }
        findCombination(ind+1,target,candidates,ans,ds);
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans= new ArrayList<>();
        List<Integer> ds=new ArrayList<>();
        findCombination(0,target,candidates,ans,ds);
        return ans;
    }
}

File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    void combination(vector<int> & candidates,int target,vector<vector<int>>& ans,vector<int>& temp,int i){
        
        int sum=accumulate(temp.begin(),temp.end(),0);
        if(i==candidates.size()){
            if(sum==target){
                ans.push_back(temp);
            }
            return;
        }
        
        if(sum<=target){
            temp.push_back(candidates[i]);
            combination(candidates,target,ans,temp,i);
            temp.pop_back();
        }
        combination(candidates,target,ans,temp,i+1);        

    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> temp;
        combination(candidates,target,ans,temp,0);
        return ans;
    }
};